test/Database/LSMTree/Model/Session.hs

```diff
@@ -266,12 +271,31 @@ instance Show Err where
       ErrBlobRefInvalidated ->
         showString "ErrBlobRefInvalidated"
       ErrCursorClosed ->
-        showString "ErrCursorCosed"
-      ErrFsError s ->
+        showString "ErrCursorClosed"
+      ErrDiskFault s ->
+        showParen (d > appPrec) $
+        showString "ErrDiskFault " .
+        showParen True (showString s)
+      ErrOther s ->
         showParen (d > appPrec) $
-        showString "ErrFsError " .
+        showString "ErrOther " .
         showParen True (showString s)
 
+-- Approximate equality for errors.
+--
+-- This approximate equality satisfies the __Reflexivity__, __Symmetry__, and
+-- __Negation__ laws for the 'Eq' class, but not __Transitivity__ and
+-- __Substitutivity__.
+--
+-- These laws should be sufficient for the purpose of comparing an error
+-- returned from the SUT to an error returned from the model.
+--
+-- Where @'Eq' 'Err'@ deviates from a typical instance is that 'ErrDiskFault' is
+-- equal to all other errors. This is important for comparing SUT exceptions to
+-- model exceptions. The model only knows /something/ went wrong, but the SUT
+-- will probably have a very specific error that it throws. In such cases where
+-- the SUT error is more specific then the model error, the errors are
+-- considered equal.
```

Comment:
This unlawful instance seems quite hacky. Is using a plain function difficult because `quickcheck-lockstep` asks for an `Eq` instance?

Reply:
The current approach piggybacks on the `quickcheck-lockstep` framework, so that we don't have to write a custom `postcondition`. 

The default postcondition is: `Either Err a == Either Err a` for some `a` that was the result of action. But with injected errors, we need approximate equality instead of syntactic equality, so I thought instead of augmenting the `postcondition` I'd just make `Eq Err` approximate.


'll try to see what it looks like if I use a plain function

Reply:
So, using a plain function is not great because we'd have to rewrite the `quickcheck-lockstep` postcondition, but it would be doable. 

However, while trying out the plain function approach I re-discovered that we had used the `Eq (Obs h a)` in the past to allow disagreements between the model and SUT, instead of the `Eq Err` instance. In particular, we had already encoded in `Eq (Obs h a)` that the model invalidates blob references conservatively, while the SUT may keep them valid for a while longer. So, I moved the check into `Eq (Obs h a)` and that makes it all simpler

Reply:
We could argue that `quickcheck-lockstep` should require just a comparison function, instead of requiring `Eq (Obs h a)`. But having a slightly unlawful `Eq (Obs h a)` achieves the same